### PR TITLE
Allow using Hash::extract() with objects implementing ArrayAccess.

### PR DESCRIPTION
**Pull request by ADmad**

### CODE REVIEW

**Comment by lorenzo** [comment]
I'm slightly concerned about this line https://github.com/cakephp/cakephp/pull/7483/files#diff-8b0f4da31563194a4b4f024bb4646c56R244

**Comment by ADmad** [comment]
@lorenzo

```
Psy Shell v0.5.2 (PHP 5.5.9-1ubuntu4.12 — cli) by Justin Hileman
>>> $a = new \ArrayObject(['foo' => null]);
=> ArrayObject {#217
     flag::STD_PROP_LIST: false,
     flag::ARRAY_AS_PROPS: false,
     iteratorClass: "ArrayIterator",
     storage: [
       "foo" => null,
     ],
   }
>>> isset($a['foo']) || array_key_exists('foo', $a)
=> true
>>> isset($a['nope']) || array_key_exists('nope', $a)
=> false
>>> 
```

File: tests/TestCase/Utility/HashTest.php [hunk 1]
```diff
@@ -149,6 +149,138 @@ public static function articleData()
...
+            ])
+        ]);
+    }
+
+    /**
+     * Data provider
+     *
+     * @return array
+     */
+    public static function articleDataSets()
```
Comment (markstory) [inline tests/TestCase/Utility/HashTest.php]: Fancy :+1:

File: src/Utility/Hash.php [hunk 2]
```diff
@@ -161,7 +161,9 @@ public static function extract(array $data, $path)
             if ($conditions) {
                 $filter = [];
                 foreach ($next as $item) {
-                    if (is_array($item) && static::_matches($item, $conditions)) {
+                    if ((is_array($item) || $item instanceof ArrayAccess) &&
+                        static::_matches($item, $conditions)
```
Comment (markstory) [inline src/Utility/Hash.php]: Should we raise an exception if `$data` is not an array or ArrayAccess?

Comment (ADmad) [inline src/Utility/Hash.php]: Well earlier also no exception was raised if value wasn't an array. It just continued on to check the next item.

Comment (ADmad) [inline src/Utility/Hash.php]: I have added data type check for argument initially passed to Hash::extract()

Comment (markstory) [inline src/Utility/Hash.php]: Thanks, I was mostly concerned with the initial parameter. Internal traversal is always a grab bag of data.

**Comment by ADmad** [comment]
The failing builds are due to segfault on 5.6 :(

**Comment by markstory** [comment]
Looks good to me.

**Comment by HavokInspiration** [comment]
I have restarted the build. They did not segfault-ed this time.

**Comment by vaughany** [comment]
@ADmad is the segfaulting related to #7299? PHP 5.6 seems to be a common factor.

**Comment by ADmad** [comment]
@vaughany I don't know. The segfault was gone when the builds were re-run.

**Pull request merged**